src/meta/src/rpc/service/notification_service.rs

```diff
@@ -71,87 +174,37 @@ where
-            self.hummock_manager
-                .pin_version(req.get_worker_id())
-                .await?;
-        }
-
-        let hummock_manager_guard = self.hummock_manager.get_read_guard().await;
+        let worker_key = WorkerKey(host_address);
 
-        let cluster_guard = self.cluster_manager.get_cluster_core_guard().await;
-        let nodes = cluster_guard.list_worker_node(WorkerType::ComputeNode, Some(Running));
-
-        match subscribe_type {
-            SubscribeType::Compactor | SubscribeType::Hummock => {
-                tables.extend(creating_tables);
-            }
-            _ => {}
-        }
+        let (tx, rx) = mpsc::unbounded_channel();
+        self.env
+            .notification_manager()
+            .insert_sender(subscribe_type, worker_key.clone(), tx)
+            .await;
 
-        // Send the snapshot on subscription. After that we will send only updates.
-        // If we let `HummockVersion` be in `meta_snapshot`, we should call `pin_version` above.
         let meta_snapshot = match subscribe_type {
-            SubscribeType::Frontend => MetaSnapshot {
-                nodes,
-                databases,
-                schemas,
-                sources,
-                sinks,
-                tables,
-                indexes,
-                users,
-                hummock_version: None,
-                parallel_unit_mappings,
-                hummock_snapshot,
-                views,
-            },
-
-            SubscribeType::Compactor => MetaSnapshot {
-                tables,
-                ..Default::default()
-            },
-
-            SubscribeType::Hummock => MetaSnapshot {
-                tables,
-                hummock_version: Some(hummock_manager_guard.current_version.clone()),
-                ..Default::default()
-            },
-
-            _ => unreachable!(),
+            SubscribeType::Compactor => self.compactor_subscribe().await,
+            SubscribeType::Frontend => {
+                self.hummock_manager
+                    .pin_snapshot(req.get_worker_id())
+                    .await?;
+                self.frontend_subscribe().await
+            }
+            SubscribeType::Hummock => {
+                self.hummock_manager
+                    .pin_version(req.get_worker_id())
+                    .await?;
+                self.hummock_subscribe().await
+            }
+            SubscribeType::Unspecified => unreachable!(),
         };
 
-        tx.send(Ok(SubscribeResponse {
-            status: None,
-            operation: Operation::Snapshot as i32,
-            info: Some(Info::Snapshot(meta_snapshot)),
-            version: self.env.notification_manager().current_version().await,
-        }))
-        .unwrap();
-
         self.env
             .notification_manager()
-            .insert_sender(subscribe_type, WorkerKey(host_address), tx)
+            .notify_snapshot(worker_key, meta_snapshot)
```

Comment:
Since `UnboundedSender` is clone-able, can we just simply send the snapshot directly here? Then we don't have to introduce `Target` in notification manager, and IIUC `notify_snapshot` will increase the global version, which does not make sense.

Reply:
Suppose that we send notification asynchronously before we get meta snapshot. The notification may be sent later than meta snapshot.

Reply:
I see. But I think we still don't need to increase the global version when notifying snapshot.

Reply:
We can increase version in `NotificationManager::notify`. Then sending notification asynchronously or sending meta snapshot will not increase notification version.  And then it's natural for us to send every kind of notification asynchronously because we don't need `callback_rx` to wait notification version. Before I wanted to reduce the modification in this PR so I withdrew these features, now I think in order to ensure that the version is not increased, we can add these features back.